### ISSUE TITLE
Support XML templates, HTML5 output

### ISSUE THREAD

**Issue by mjog** [thymeleaf/thymeleaf#777]
I am in the process of converting a Thymeleaf 2.1 installation to 3.0 that has a substantial number of existing templates as XML files that contain HTML5 markup.

XML is used as the template format since it is trivial to author, parse and detect errors such as missing end tags, etc both during editing, development and via CI, using existing tools. Templates have standard HTML5 doctype declarations and ".xml" file name extensions.

HTML5 is the desired output format because it is the only sane choice. The rendered output is sent as `text/html`.

This seemed to work better with Thymeleaf 2.1 since I could just set the format to be HTML5, however this does not work with 3.0 - an XML declaration is output and this breaks rendering in browsers.

Thus it would be useful to decouple the input and output formats. Thymeleaf could (should?) detect that the template format specifies HTML5 content and do the right thing with the output.

**Comment by garretwilson** [comment]
I'm having a similar issue. I am using Thymeleaf from Spring Boot 2.3.3.RELEASE. My templates are well-formed XHTML5 files with `.xhtml` extensions, and an XML declaration:

```xml
<?xml version="1.0" encoding="UTF-8"?>
```

I specify the suffix in `application.properties`:

```properties
spring.thymeleaf.suffix=.xhtml
```

Thymeleaf picks up the XHTML5 template files and generates views of type `text/html;charset=utf-8`. This is all correct so far.

However Thymeleaf is leaving the XML declaration (`<?xml version="1.0" encoding="UTF-8"?>`). In the generated HTML files. The XML declaration should be removed; it is not appropriate for content type `text/html`.

As the XML declaration is optional for XML files, I will remove them from the XHTML5 source as a workaround. Still Thymeleaf should already know this and just remove them.

**Comment by misyntropy** [comment]
I'm new here, but I am a little confused by this issue - since thymeleaf is an XML based templating engine, surely serving pages as `application/xhtml+xml` is an easier option right? What use case makes `text/html` better?

**Comment by garretwilson** [comment]
> … surely serving pages as `application/xhtml+xml` is an easier option right? What use case makes `text/html` better?

It's a very long and complicated story. Originally the W3C was going to go a completely XML route, and had a whole XHTML 2.0 initiative, with modules that could be plugged in and DTDs and stuff; that effort basically stalled and died for many reasons, not least of which was the inability of the specification process to be quick, nimble, and flexible. The (`application/xhtml+xml` media type was chosen to use `application/` rather than `text/` because of an old MIME rule that was interpreted to mean that any `text/*` media type had to default to the `US-ASCII` charset. Luckily more rational thoughts prevailed, which is why we can have `text/*` media types default to `UTF-8` nowadays, but I'm getting sidetracked in the story.)

In any case, `application/xhtml+xml` indicates among other things that the browser should at least handle namespaces correctly, and the Microsoft browsers could never figure out how to implement XML namespaces (even though they had existed for well over a decade), so Internet Explorer would just die if it encountered an HTML file marked as `application/xhtml+xml`. There were possible messy workaround both on the client side and on the server side for detecting IE and pretending the content was `text/html`, but in the end people just went with `text/html`.

In the meantime the WHATWG succeeded in putting out a new version of HTML since the W3C just couldn't get its act together with XHTML. The new HTML (otherwise known as HTML5) supports being encoded as `application/xhtml+xml`, and that's nice for editing, but you still wouldn't want to serve it as that because I wouldn't want to start the pain of testing browsers to see if they support namespaces correctly now, and I would guess that Microsoft never got around to fixing their support of this MIME type (I haven't checked because nobody cares anymore), and I haven't heard of any libraries that pay much mind to namespaces.

To summarize: Any hope of `application/xhtml+xml` working in a browser and served from the Internet died well over a decade ago because Microsoft browsers croaked when they encountered such content, so everybody just uses `text/html`.

**Comment by misyntropy** [comment]
Oh sorry — I did not make myself clear — I learned of this long, sordid story in great detail, in parallel with really understanding the 2010's HTML/CSS/JS web stack (I played around with it in high school but that doesn't count) and how this played out painfully, with murder after murder of XML-related technology and feature. That aside about `text/*` being interpreted to require a default of `US-ASCII` encoding was new to me though, thanks.

In particular, while Firefox was supporting XHTML by 2003, IE didn't support XHTML till 2011, after it was too late and WHATWG had done its thing. But ever since Internet Explorer 9, IE has been doing XHTML properly (I tried). JS libraries on the browser that are namespace-blind will simply stick only to elements in the HTML namespace since it's been made the default while running scripts.

Ironically, although XHTML is in "maintenance mode" in every browser today, it is actually better supported now than in its heydays, go figure. If you're not bothered about IE<=8, and since the XML serialisation of HTML5 still exists and works well, maybe just use it?

**Comment by garretwilson** [comment]
> Ironically, although XHTML is in "maintenance mode" in every browser today, it is actually better supported now than in its heydays, go figure.

Interesting. I hadn't even considered touching it. But realistically, can you point me to a single site on the Internet serving HTML files as `application/xhtml+xml` for production, but more than a simple "hello world" page? I'm very curious now.

Do you really think modern libraries such as React.js would work when served as `application/xhtml+xml`? I have no experience to go on, but I shudder to think of even trying it.

**Comment by misyntropy** [comment]
[The one and only...](https://golem.ph.utexas.edu/~distler/blog/). I kid though, there are others (edit: added [this one](https://www.devever.net/~hl/xhtml2) ), but of course rare, as you would already know.

I don't have any experience with react.js, but I'm guessing it will, considering that once the DOM is built, it's indistinguishable to scripts as to what serialisation the markup came in.

**Comment by garretwilson** [comment]
> … once the DOM is built, it's indistinguishable to scripts as to what serialisation the markup came in …

… except that there are two sets of DOM methods (or at least there used to be): one set of which is not namespace aware; and another which is namespace aware, with which you have to provide namespaces for what you're trying to access. Last I checked these different methods behave in subtly different ways.

Maybe everything "just works" now, 20 years after I abandoned serving `application/xhtml+xml`, but since almost nobody actually uses that feature, I would be willing to bet there are all sorts of latent bugs across browsers that nobody finds. I would also guess that many frameworks were written quick-and-dirty with no thought of namespaces. I shudder again just thinking of going through that again: trying to make sites work across browsers, until finally enough namespace-related bugs crop up that I have to throw in the towel and convert everything back. I have enough problems to work around, so for me right now I'm just going to serve things as `text/html` and let it be.

**Comment by garretwilson** [comment]
> … there are others …

I consider the sites you mentioned pretty much "hello world" pages, with little if any server communication and significant dynamic DOM updates. Nothing against those sites, except that that they provide no real indication of whether any "interesting" functionality on the web actually works with modern libraries.

**Comment by misyntropy** [comment]
You're more than justified, as are all ordinary web authors, in sticking to `text/html` given the witch-hunt XML has been subject to. The discovery of latent bugs is a possibility, that can be quite annoying, in particular. But the namespace-awareness-of-JS question is [rather well answered](https://stackoverflow.com/questions/23319537/html-5-inline-svg-and-namespace-awareness-for-svg-dom) — namespace-unaware DOM methods automatically target the HTML namespace. 

I lol'ed at "Hello World" pages, but I got it.  They lack dynamic DOM updates and are basically web pages/documents, not web apps. Maybe you can just try serving as XHTML (only MIME type change and nothing else), and see how bad it is? Just a thought.... I thought Thymeleaf would be the one piece of software whose users would prefer `application/xhtml+xml`, that's all.

**Comment by garretwilson** [comment]
> I lol'ed at "Hello World" pages …

I didn't mean any disrespect at them, and I imagine a lot of work and sweat and maybe even tears went into some of them. I probably should have said "static pages", which is closer to what I was referring to. As you surmised, the issue is whether they are the type of pages that would reveal latent bugs with browser namespace handling through intense DOM manipulation.

> … namespace-unaware DOM methods automatically target the HTML namespace.

Um, I don't think so. You might want to double-check on that. Namespace-unaware DOM methods, if I remember correctly, do their lookup based upon the qnames (the prefix+localName qualified names), without looking at namespaces at all. Thus you could do something in your JavaScript if you see the element `<p>`, even though in reality the default namespace might have been changed and `<p>` is not in the HTML namespace anymore and refers to some other `<p>` in a different namespace (something that might happen if a framework uses the non-namespace aware method but the underlying document is served as `application/xhtml+xml`).

The [Stack Overflow question you mentioned](https://stackoverflow.com/q/23319537) actually talks about the other side of the coin: a document served as `text/html` will have its namespaces automatically placed in the `http://www.w3.org/1999/xhtml` namespace. Thus a _namespace-aware_ method would succeed in finding HTML elements, even though the file was not served as `application/xhtml+xml`. (On a related note, you might be interested in jhy/jsoup#1837.)

But that's the opposite situation of what we were discussing; it doesn't get us anywhere for a document served as `application/xhtml+xml` but the JavaScript accesses the DOM methods that are _not_ namespace unaware.

**Comment by misyntropy** [comment]
Good point, [this question](https://stackoverflow.com/questions/24343331/how-to-get-an-element-by-id-placed-in-some-namespace-in-xhtml-via-js) sheds some light on it. I also just tried it out with an XHTML page — getElementById searches among all tags anyway, getElementsByTagName searches on the XHTML namespace by default. Attributes are anyway in the null namespace by default, even if their elements are in different NSes. There's nothing you can do with namespace-unaware methods to cause any other namespace-unaware methods to do the wrong thing — the default namespace is assumed always to be the XHTML namespace.

**Comment by mjog** [comment]
> I'm new here, but I am a little confused by this issue - since thymeleaf is an XML based templating engine, surely serving pages as `application/xhtml+xml` is an easier option right? What use case makes `text/html` better?

The salient points here are (quoting myself in the OP):

> XML is used as the template format since it is trivial to author, parse and detect errors such as missing end tags, etc both during editing, development and via CI, using existing tools

and:

> HTML5 is the desired output format because it is the only sane choice

TL;DR: XML templates avoids tag soup, HTML5 is only thing supported in browsers going forward

**Comment by mjog** [comment]
Please don't discuss HTML vs XHTML and browser support (or lack thereof) for XHTML here, that is OT.

**Comment by misyntropy** [comment]
@mjog XML-syntax for HTML5 ("XHTML5") is baked into the HTML5 spec, and every major browser supports it. Even going forward, its only things like XSLT or XPath that might be dropped (Only Chrome has looked into it), not the XML parser itself. I don't understand what you mean by 

> HTML5 is only thing supported in browsers going forward

**Comment by mjog** [comment]
> Please don't discuss HTML vs XHTML and browser support (or lack thereof) for XHTML here, that is OT.

**Comment by danielfernandez** [comment]
Closing this (housekeeping).

**Comment by mjog** [comment]
Oh! Has it been fixed?
